base/sequence_clock.go

```diff
@@ -462,3 +480,219 @@ func GetMinimumClock(a SequenceClock, b SequenceClock) *SequenceClockImpl {
 	}
 	return minClock
 }
+
+// IndexablePartitionClock is used to persist clocks as JSON.  Used for rollback view query.
+type IndexablePartitionClock struct {
+	Cas            uint64
+	Key            string
+	PartitionClock PartitionClock
+	storage        *IndexablePartitionClockStorage
+}
+
+func NewIndexablePartitionClock(key string, channelName string) *IndexablePartitionClock {
+	return &IndexablePartitionClock{
+		Key:            key,
+		PartitionClock: make(PartitionClock),
+		storage: &IndexablePartitionClockStorage{
+			VbNos:   make([]uint16, 0),
+			Seqs:    make([]uint64, 0),
+			Channel: channelName,
+		},
+	}
+}
+
+// Updates clock from another clock.  Will not lower sequence numbers (in case of concurrent writers) - logs warning in this scenario.
+func (s *IndexablePartitionClock) Update(clock PartitionClock, allowRollback bool) (changed bool) {
+	for vb, seq := range clock {
+		currentSequence := s.PartitionClock.GetSequence(vb)
+		if seq > currentSequence || (seq < currentSequence && allowRollback) {
+			s.PartitionClock.SetSequence(vb, seq)
+			changed = true
+		} else if seq < currentSequence {
+			Warn("Ignored update of sequence clock for vb:[%d] existing:[%d] update:[%d]", vb, currentSequence, seq)
+		}
+	}
+	return changed
+}
+
+type IndexablePartitionClockStorage struct {
+	Channel string   `json:"channel"`
+	VbNos   []uint16 `json:"vb"`
+	Seqs    []uint64 `json:"seq"`
+}
+
+func (s *IndexablePartitionClockStorage) update(clock PartitionClock) {
+	s.VbNos = s.VbNos[:0]
+	s.Seqs = s.Seqs[:0]
+	for vb, seq := range clock {
+		s.VbNos = append(s.VbNos, vb)
+		s.Seqs = append(s.Seqs, seq)
+	}
+}
+
+func (i *IndexablePartitionClock) MarshalJSON() ([]byte, error) {
+	// Reset storage (without re-alloc)
+	i.storage.update(i.PartitionClock)
```

Comment:
We could introduce `i.storage.reset()` which calls down to `i.storage.update(i.PartitionClock)`.  That way the usage is easier to grok, but `i.storage.update()` remains easily testable.